src/configupdater/container.py

```diff
@@ -70,5 +92,20 @@ def __len__(self) -> int:
         return len(self._structure)
 
     def append(self: C, block: T) -> C:
-        self._structure.append(block)
+        self._structure.append(self._adopt_or_copy(block))
+        return self
+
+    def insert(self: C, idx: int, block: T) -> C:
+        self._structure.insert(idx, self._adopt_or_copy(block))
         return self
+
+    def _adopt_or_copy(self, block: T) -> T:
+        """Run this function before adopting any block inside the structure"""
+        if block._container == self:
+            return block
+
+        if block.is_attached():
+            warnings.warn(AlreadyAttachedWarning())
+            block = copy(block)
+
+        return block._attach(self)
```

Comment:
I am not sure here if we should warn the user, or simply do the copying implicitly behind the scenes...